File: nix-cxx/src/nix.cpp
#include "nix-cxx/include/nix.h"
#include <memory>

using namespace std;

namespace nix_cxx
{
  std::unique_ptr<nix::FlakeRef> parse_flakeref(rust::Str url)
  {
    nix::FlakeRef originalRef = nix::parseFlakeRef(string(url), nix::absPath("."));
    return std::make_unique<nix::FlakeRef>(originalRef);
  }

  rust::String flakeref_to_string(std::unique_ptr<nix::FlakeRef> flakeref)
  {
    auto s = flakeref->to_string();
    return rust::String(s);
  }
}

File: nix-cxx/src/nix.rs
#[cxx::bridge(namespace = "nix_cxx")]
pub(crate) mod ffi {
    unsafe extern "C++" {
        include!("nix-cxx/include/nix.h");

        type FlakeRef;

        fn parse_flakeref(url: &str) -> UniquePtr<FlakeRef>;
        fn flakeref_to_string(flakeref: UniquePtr<FlakeRef>) -> String;
    }
}
